### PR TITLE
UCP/API: close protocol for stream

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
This PR is a next round of https://github.com/openucx/ucx/pull/2463 which was reverted in https://github.com/openucx/ucx/pull/2635 due to incompleteness. There are a few changes:

1. The API is more stream oriented. Reason: only STREAM ops are EP to EP, TAG/RMA/AMO ops work with worker (EP to WORKER).
2. Notification about remote disconnect event replaced by event flags in out parameter of ucp_stream_worker_poll routine instead of error handling callback before. It makes API closer to socket semantic.

Backward compatibility note:
 - The change adds new ep closure mode with remote notification with
   ucp_stream_poll_flags_t::UCP_STREAM_POLL_FLAG_CLOSED
 - Added new flag ucp_stream_poll_flags_t::UCP_STREAM_POLL_FLAG_IN to
   make user abale to diverse the events on the polled-out endpoint
 - Adding ucp_stream_poll_flags_t does not break compatibility
   because before this change user should close the EPs according
   his protocol or application behavior without any notification
   like UCP_STREAM_POLL_FLAG_CLOSED. If user does not use
   UCP_EP_CLOSE_MODE_STREAM_CLOSE, he can ignore ucp_stream_poll_flags_t
   and handle only data to receive

@yosefe

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7219/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -251,9 +251,28 @@ enum ucp_ep_close_mode {
                                               for all endpoints created on
                                               both (local and remote) sides to
                                               avoid undefined behavior. */
-    UCP_EP_CLOSE_MODE_FLUSH         = 1  /**< @ref ucp_ep_close_nb schedules
+    UCP_EP_CLOSE_MODE_FLUSH         = 1, /**< @ref ucp_ep_close_nb schedules
                                               flushes on all outstanding
                                               operations. */
+    UCP_EP_CLOSE_MODE_STREAM_SYNC   = 2  /**< @ref ucp_ep_close_nb will flush
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: IMHO STREAM_SYNC is not a good name , it does not indicate that the stream is closed/disconnected
maybe STREAM_END / STREAM_FIN ?

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -251,9 +251,28 @@ enum ucp_ep_close_mode {
...
+                                              the local endpoint and handle the
+                                              remote endpoint by generating an
+                                              close event
+                                              @ref UCP_STREAM_POLL_FLAG_NVAL on
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: "NVAL" is not a word, need better name

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -251,9 +251,28 @@ enum ucp_ep_close_mode {
                                               for all endpoints created on
                                               both (local and remote) sides to
                                               avoid undefined behavior. */
-    UCP_EP_CLOSE_MODE_FLUSH         = 1  /**< @ref ucp_ep_close_nb schedules
+    UCP_EP_CLOSE_MODE_FLUSH         = 1, /**< @ref ucp_ep_close_nb schedules
                                               flushes on all outstanding
                                               operations. */
+    UCP_EP_CLOSE_MODE_STREAM_SYNC   = 2  /**< @ref ucp_ep_close_nb will flush
+                                              the local endpoint and handle the
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: Need to reword, maybe something like this:
`@ref ucp_ep_close_nb will close the stream connection. The remote side will have a @ref UCP_STREAM_POLL_FLAG_CLOSED flag for the remote endpoint if @ref ucp_stream_worker_poll is called on the remote worker...`

File: src/ucp/api/ucp.h [hunk 4]
```diff
@@ -409,6 +428,30 @@ typedef enum {
...
+ *
+ * This enumeration defines types of events on a stream associated endpoint.
+ * These flags will be set in the field @ref ucp_stream_poll_ep_t::flags of out
+ * parameter @a poll_eps returned by @ref ucp_stream_worker_poll routine.
+ */
+typedef enum {
+    /**
+     * There is data to receive.
+     */
+    UCP_STREAM_POLL_FLAG_IN   = UCS_BIT(0),
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: this breaks compatibility

Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h]: will document in commit message and PR desc

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7250/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
Not related failure, opened an issue https://github.com/openucx/ucx/issues/2706
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7252/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
@yosefe pls take a look

File: src/ucp/api/ucp.h [hunk 5]
```diff
@@ -409,6 +423,33 @@ typedef enum {
...
+     */
+    UCP_STREAM_POLL_FLAG_IN     = UCS_BIT(0),
+
+    /**
+     * The stream connection closed by remote peer. If the flag set together
+     * with @ref UCP_STREAM_POLL_FLAG_IN, the data can be received by @ref
+     * ucp_stream_recv_nb or @ref ucp_stream_recv_data_nb before closing the
+     * local endpoint by @ref ucp_ep_close_nb. The local endpoint should be
+     * closed in @ref UCP_EP_CLOSE_MODE_STREAM_CLOSE mode to avoid protocol
+     * mismatch, excepting remote failure cases where @ref
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: "protocol mismatch" is too generic, what does it mean?

File: src/ucp/api/ucp.h [hunk 6]
```diff
@@ -1869,7 +1911,7 @@ typedef enum ucp_mem_advice {
     UCP_MADV_NORMAL   = 0,  /**< No special treatment */
     UCP_MADV_WILLNEED       /**< can be used on the memory mapped with
                                  @ref UCP_MEM_MAP_NONBLOCK to speed up memory
-                                 mapping and to avoid page faults when 
+                                 mapping and to avoid page faults when
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: i would put all whitespace fixes in another PR, since this is API

File: src/ucs/type/status.h [hunk 7]
```diff
@@ -70,6 +70,7 @@ typedef enum {
     UCS_ERR_TIMED_OUT              = -20,
     UCS_ERR_EXCEEDS_LIMIT          = -21,
     UCS_ERR_UNSUPPORTED            = -22,
+    UCS_ERR_REMOTE_DISCONNECT      = -24,
```
Comment (yosefe) [inline src/ucs/type/status.h]: how is this related to the PR? i don't see this error code in the documentation added in this PR..

Comment (evgeny-leksikov) [reply src/ucs/type/status.h]: This error code will be returned by ucp_stream_send_nb if the endpoint closed by remote peer. I added this change to this PR because status.h is installable header. Description to ucp_stream_send_nb won't be changed because we don't describe all error codes to API routines.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7530/ for details (Mellanox internal link).

File: src/ucs/type/status.h (hunk 7 above)
Comment (yosefe) [reply src/ucs/type/status.h]: so let's document that closing a stream with close protocol would generate such error on remote side

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7536/ for details (Mellanox internal link).

File: src/ucs/type/status.h (hunk 7 above)
Comment (evgeny-leksikov) [reply src/ucs/type/status.h]: done

**Comment by yosefe** [comment]
@shamisp @gmegan pls take a look

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7542/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
👍 @bbenton has to approve this.

**Comment by shamisp** [comment]
Obviously the conflict has to be resolved before merge

File: src/ucp/api/ucp.h [hunk 8]
```diff
@@ -251,9 +251,25 @@ enum ucp_ep_close_mode {
...
+                                              on the remote worker. This event
+                                              puts the remote endpoint into an
+                                              error state, all following @ref
+                                              ucp_stream_send_nb calls on the
+                                              remote endpoint will return @ref
+                                              UCS_ERR_REMOTE_DISCONNECT error
+                                              code. The UCP should be
+                                              initialized with @ref
+                                              UCP_FEATURE_STREAM, otherwise
+                                              using of stream API is invalid. */
```
Comment (bbenton) [inline src/ucp/api/ucp.h]: I suggest the following re-wording for clarity:
`@ref` ucp_ep_close_nb will close the stream connection. The remote side will return `@ref` UCP_STREAM_POLL_FLAG_CLOSED for the remote endpoint if `@ref` ucp_stream_worker_poll is called on the remote worker. This mode puts the remote endpoint into an error state. All subsequent `@ref` ucp_stream_send_nb calls on the remote endpoint will return `@ref` UCS_ERR_REMOTE_DISCONNECT. 
`@note` To use the stream API, UCP should be initialized with `@ref` UCP_FEATURE_STREAM, otherwise the use of the stream API is invalid.

File: src/ucp/api/ucp.h [hunk 9]
```diff
@@ -409,6 +425,33 @@ typedef enum {
...
 
 
+/**
+ * @ingroup UCP_COMM
+ * @brief flags to indicate types of events on a stream associated endpoint.
+ *
+ * This enumeration defines types of events on a stream associated endpoint.
+ * These flags will be set in the field @ref ucp_stream_poll_ep_t::flags of out
+ * parameter @a poll_eps returned by @ref ucp_stream_worker_poll routine.
+ */
```
Comment (bbenton) [inline src/ucp/api/ucp.h]: suggested re-wording:This enumeration defines the possible events that can be raised on a stream associated endpoint. These flags will be set in the `@ref` ucp_stream_poll_ep_t::flags field of the `@a` poll_eps array returned by the `@ref` ucp_stream_worker_poll routine.

File: src/ucp/api/ucp.h [hunk 10]
```diff
@@ -409,6 +425,33 @@ typedef enum {
...
+    UCP_STREAM_POLL_FLAG_IN     = UCS_BIT(0),
+
+    /**
+     * Remote peer closed the stream connection by @ref ucp_ep_close_nb in @ref
+     * UCP_EP_CLOSE_MODE_STREAM_CLOSE mode. If the flag set together
+     * with @ref UCP_STREAM_POLL_FLAG_IN, the data can be received by @ref
+     * ucp_stream_recv_nb or @ref ucp_stream_recv_data_nb before closing the
+     * local endpoint by @ref ucp_ep_close_nb. The local endpoint should be
+     * closed in @ref UCP_EP_CLOSE_MODE_STREAM_CLOSE mode too, excepting remote
+     * failure cases where @ref UCP_EP_CLOSE_MODE_FORCE may be used.
```
Comment (bbenton) [inline src/ucp/api/ucp.h]: More suggested wording changes:

The remote peer closed the stream connection by `@ref` ucp_ep_close_nb in `@ref` UCP_EP_CLOSE_MODE_STREAM_CLOSE mode. If `@ref` UCP_STREAM_POLL_FLAG_IN is also set, the associated data can be received by calling either `@ref` ucp_stream_recv_nb or `@ref` ucp_stream_recv_data_nb before closing the local endpoint by `@ref` ucp_ep_close_nb. The local endpoint should be closed in `@ref` UCP_EP_CLOSE_MODE_STREAM_CLOSE mode too, except in remote failure cases where `@ref` UCP_EP_CLOSE_MODE_FORCE may be used.
